### DISCUSSION TITLE
Sparse node features

### DISCUSSION THREAD

**Discussion by Rajjaa** [pyg-team/pytorch_geometric#4661] [Q&A] [answered]
Hello,

I would like to know if there is a way to develop GNN models that accept sparse node features?

Thanks

**Comment by rusty1s** [comment, marked as answer]
Some GNNs support sparse node feature matrices directly as input such as `GCNConv`, but this is sadly not possible for all. For others, it might be a good idea to first project the sparse features into a low-dimensional (but dense) embedding space, and use that as input into the GNN:
```python
lin = Linear(num_features, 32)
x = lin(sparse_x)
```

**Reply by Rajjaa** [reply]
Thank you for your reply but the two scenarios you mentioned throw a TypeError when using sparse node features from the NELL dataset:
`TypeError: linear(): argument 'input' (position 1) must be Tensor, not SparseTensor`
It seems that the Linear layer doesn't accept sparse tensors.

**Reply by rusty1s** [reply]
You will need to use the `torch.sparse_coo_tensor` from PyTorch for this, sorry about the confusion:
```python
lin(sparse_x.to_torch_sparse_coo_tensor())
```

**Reply by Rajjaa** [reply]
Thank you it is working now. So why there is such a difference between sparse tensors from PyTorch and sparse tensors from `torch-sparse`?
